src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -6243,5 +6155,162 @@ public static void Main()
+        System.Console.WriteLine(c.x + "" "" + c.y);
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "1 hello", additionalRefs: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("C..ctor(int, string)", @"
+{
+  // Code size       26 (0x1a)
+  .maxstack  2
+  .locals init (long V_0,
+                string V_1)
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""object..ctor()""
+  IL_0006:  ldarg.1
+  IL_0007:  conv.i8
+  IL_0008:  stloc.0
+  IL_0009:  ldarg.2
+  IL_000a:  stloc.1
+  IL_000b:  ldarg.0
+  IL_000c:  ldloc.0
+  IL_000d:  stfld      ""long C.x""
+  IL_0012:  ldarg.0
+  IL_0013:  ldloc.1
+  IL_0014:  stfld      ""string C.y""
+  IL_0019:  ret
+}");
+        }
+
+        [Fact]
+        public void DeconstructAssignInConstructor()
+        {
+            string source = @"
+public class C
+{
+    public long x;
+    public string y;
+
+    public C(C oldC) => (x, y) = oldC;
+    public C() { }
+
+    public void Deconstruct(out int a, out string b)
+    {
+        a = 1;
+        b = ""hello"";
+    }
+
+    public static void Main()
+    {
+        var oldC = new C() { x = 1, y = ""hello"" };
+        var newC = new C(oldC);
+        System.Console.WriteLine(newC.x + "" "" + newC.y);
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "1 hello", additionalRefs: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("C..ctor(C)", @"
+{
+  // Code size       34 (0x22)
+  .maxstack  3
+  .locals init (int V_0,
+                string V_1,
+                long V_2)
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""object..ctor()""
+  IL_0006:  ldarg.1
+  IL_0007:  ldloca.s   V_0
+  IL_0009:  ldloca.s   V_1
+  IL_000b:  callvirt   ""void C.Deconstruct(out int, out string)""
+  IL_0010:  ldloc.0
+  IL_0011:  conv.i8
+  IL_0012:  stloc.2
+  IL_0013:  ldarg.0
+  IL_0014:  ldloc.2
+  IL_0015:  stfld      ""long C.x""
+  IL_001a:  ldarg.0
+  IL_001b:  ldloc.1
+  IL_001c:  stfld      ""string C.y""
+  IL_0021:  ret
+}");
+        }
+
+        [Fact]
+        public void AssignInConstructorWithProperties()
+        {
+            string source = @"
+public class C
+{
+    public long X { get; set; }
+    public string Y { get; }
+    private int z;
+    public ref int Z { get { return ref z; } }
+
+    public C(int a, string b, ref int c) => (X, Y, Z) = (a, b, c);
+
+    public static void Main()
+    {
+        int number = 2;
+        var c = new C(1, ""hello"", ref number);
+        System.Console.WriteLine($""{c.X} {c.Y} {c.Z}"");
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "1 hello 2", additionalRefs: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("C..ctor(int, string, ref int)", @"
+{
+  // Code size       39 (0x27)
+  .maxstack  4
+  .locals init (long V_0,
+                string V_1,
+                int V_2,
+                long V_3)
```

Comment:
There is room for further optimization. I don't think we need `V_1` or `V_3` here.